File: app/src/main/java/adapters/CommentAdapter.java
package adapters;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import androidx.annotation.NonNull;
import com.google.android.material.snackbar.Snackbar;
import androidx.recyclerview.widget.RecyclerView;

import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.signature.ObjectKey;
import com.firebase.ui.firestore.FirestoreRecyclerAdapter;
import com.firebase.ui.firestore.FirestoreRecyclerOptions;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.Transaction;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.sqube.tipshub.LoginActivity;
import com.sqube.tipshub.MemberProfileActivity;
import com.sqube.tipshub.MyProfileActivity;
import com.sqube.tipshub.R;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import de.hdodenhof.circleimageview.CircleImageView;
import models.Comment;
import models.Post;
import models.SnapId;
import models.UserNetwork;
import services.GlideApp;
import utils.Calculations;
import utils.FirebaseUtil;
import utils.Reusable;

import static utils.Reusable.getPlaceholderImage;

public class CommentAdapter extends FirestoreRecyclerAdapter<Comment, CommentAdapter.CommentHolder>{
    private final String TAG = "CommentAdapter";
    private Activity activity;
    private Context context;
    private String userId;
    private String mainPostId;
    private Calculations calculations;
    private StorageReference storageReference;
    private ArrayList<SnapId> repliedList = new ArrayList<>();
    private boolean anchorSnackbar;

    public ArrayList<SnapId> getRepliedList() {
        return repliedList;
    }

    public void setUserId(String userId) {
        this.userId = userId;
    }

    public void resetRepliesList(){
        repliedList.clear();
    }

    public CommentAdapter(String mainPostId, Query query, String userID, Activity activity, Context context, boolean anchorSnackbar) {
        /*
        Configure recycler adapter options:
        query defines the request made to Firestore
        Comment.class instructs the adapter to convert each DocumentSnapshot to a Comment object
        */
        super(new FirestoreRecyclerOptions.Builder<Comment>()
                .setQuery(query, Comment.class)
                .build());
        this.anchorSnackbar = anchorSnackbar;

        Log.i(TAG, "CommentAdapter: created");
        this.activity = activity;
        this.context = context;
        calculations = new Calculations(context);
        this.setUserId(userID);
        this.mainPostId = mainPostId;
        storageReference = FirebaseStorage.getInstance().getReference()
                .child("profile_images");
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    protected void onBindViewHolder(@NonNull CommentHolder holder, final int position, @NonNull final Comment model) {
        Log.i(TAG, "onBindViewHolder: executed");
        final TextView mComment = holder.mComment;
        final TextView mUsername = holder.mUsername;
        final TextView mTime = holder.mTime;
        final TextView mLikesCount = holder.mLikes;
        final TextView mDislikesCount = holder.mDislikes;
        final CircleImageView imgDp = holder.imgDp;
        final ImageView imgLikes = holder.imgLikes;
        final ImageView imgDislikes = holder.imgDislike;
        final ImageView imgReply = holder.imgReply;
        final ImageView imgOverflow = holder.imgOverflow;
        final String postId = getSnapshots().getSnapshot(position).getId();

        if(model.isFlag())
            holder.lnrContainer.setBackgroundColor(context.getResources().getColor(R.color.comment_flagged));
        else
            holder.lnrContainer.setBackgroundColor(context.getResources().getColor(R.color.comment_bg));

        imgReply.setVisibility(model.getUserId().equals(userId)? View.GONE: View.VISIBLE);

        //set username and comment content
        mUsername.setText(model.getUsername());
        mComment.setText(model.getContent());
        Reusable.applyLinkfy(context, model.getContent(), mComment);
        mTime.setText(Reusable.getTime(model.getTime()));
        imgLikes.setColorFilter(model.getLikes().contains(userId)?
                context.getResources().getColor(R.color.likeGold): context.getResources().getColor(R.color.likeGrey));

        imgDislikes.setColorFilter(model.getDislikes().contains(userId)?
                context.getResources().getColor(R.color.likeGold): context.getResources().getColor(R.color.likeGrey));

        mLikesCount.setText(model.getLikesCount()==0? "":String.valueOf(model.getLikesCount()));
        mDislikesCount.setText(model.getDislikesCount()==0? "":String.valueOf(model.getDislikesCount()));

        GlideApp.with(context)
                .load(storageReference.child(model.getUserId()))
                .placeholder(R.drawable.dummy)
                .error(getPlaceholderImage(model.getUserId().charAt(0)))
                .signature(new ObjectKey(model.getUserId()+"_"+Reusable.getSignature()))
                .into(imgDp);

        //listen to dp click and open user profile
        imgDp.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(model.getUserId().equals(userId)){
                    imgDp.getContext().startActivity(new Intent(context, MyProfileActivity.class));
                }
                else{
                    Intent intent = new Intent(context, MemberProfileActivity.class);
                    intent.putExtra("userId", model.getUserId());
                    imgDp.getContext().startActivity(intent);
                }
            }
        });

        //listen to username click and open user profile
        mUsername.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(model.getUserId().equals(userId)){
                    mUsername.getContext().startActivity(new Intent(context, MyProfileActivity.class));
                }
                else{
                    Intent intent = new Intent(context, MemberProfileActivity.class);
                    intent.putExtra("userId", model.getUserId());
                    mUsername.getContext().startActivity(intent);
                }
            }
        });

        imgLikes.setOnClickListener(v -> {
            Log.i(TAG, "onClick: Key is " + postId);
            if (userId.equals(Calculations.GUEST)) {
                loginPrompt(imgLikes);
                return;
            }
            if(model.getDislikes().contains(userId)){
                holder.imgLikes.setColorFilter(context.getResources().getColor(R.color.likeGold));
                holder.imgDislike.setColorFilter(context.getResources().getColor(R.color.likeGrey));
                holder.mLikes.setText(String.valueOf(model.getLikesCount()+1));
                holder.mDislikes.setText(model.getDislikesCount()-1>0? String.valueOf(model.getDislikesCount()-1):"");
            }
            else{
                if(model.getLikes().contains(userId)){
                    holder.imgLikes.setColorFilter(context.getResources().getColor(R.color.likeGrey));
                    holder.mLikes.setText(model.getLikesCount()-1>0?String.valueOf(model.getLikesCount()-1):"");
                }
                else{
                    holder.imgLikes.setColorFilter(context.getResources().getColor(R.color.likeGold));
                    holder.mLikes.setText(String.valueOf(model.getLikesCount()+1));
                }
            }
            String substring = model.getContent().substring(0, Math.min(model.getContent().length(), 90));
            calculations.onCommentLike(userId, model.getUserId(), postId, mainPostId, substring);
        });

        imgReply.setOnClickListener(v -> {
            if (userId.equals(Calculations.GUEST)) {
                loginPrompt(imgReply);
                return;
            }
            EditText edtComment = activity.findViewById(R.id.edtComment);
            String comment = edtComment.getText().toString();
            if(comment.isEmpty())
                edtComment.setText(String.format("@%s ", model.getUsername()));
            else
                edtComment.setText(String.format("%s @%s ", comment.trim(), model.getUsername()));
            edtComment.setSelection(edtComment.getText().length());
            repliedList.add(new SnapId(model.getUserId(), model.getUsername()));
        });

        imgDislikes.setOnClickListener(v -> {
            Log.i(TAG, "onClick: Key is " + postId);
            if (userId.equals(Calculations.GUEST)) {
                loginPrompt(imgDislikes);
                return;
            }

            if(model.getLikes().contains(userId)){
                holder.imgLikes.setColorFilter(context.getResources().getColor(R.color.likeGrey));
                holder.imgDislike.setColorFilter(context.getResources().getColor(R.color.likeGold));
                holder.mLikes.setText(model.getLikesCount()-1>0? String.valueOf(model.getLikesCount()-1):"");
                holder.mDislikes.setText(String.valueOf(model.getDislikesCount()+1));
            }
            else{
                if(model.getDislikes().contains(userId)){
                    holder.imgDislike.setColorFilter(context.getResources().getColor(R.color.likeGrey));
                    holder.mDislikes.setText(model.getDislikesCount()-1>0? String.valueOf(model.getDislikesCount()-1): "");
                }
                else{
                    holder.imgDislike.setColorFilter(context.getResources().getColor(R.color.likeGold));
                    holder.mDislikes.setText(String.valueOf(model.getDislikesCount()+1));
                }
            }
            String substring = model.getContent().substring(0, Math.min(model.getContent().length(), 90));
            calculations.onCommentDislike(userId, model.getUserId(), postId, mainPostId, substring);
        });

        imgOverflow.setOnClickListener(v -> displayOverflow(model, model.getUserId(), postId, holder.imgOverflow));
    }

    private void loginPrompt(View view) {
        androidx.appcompat.app.AlertDialog.Builder builder = new androidx.appcompat.app.AlertDialog.Builder(view.getRootView().getContext(),
                R.style.CustomMaterialAlertDialog);
        builder.setMessage("You have to login first")
                .setNegativeButton("Cancel", (dialogInterface, i) -> {})
                .setPositiveButton("Login", (dialogInterface, i) -> view.getContext().startActivity(new Intent(view.getContext(), LoginActivity.class)))
                .show();
    }

    private void displayOverflow(Comment model, String commentUserId, String postId, ImageView imgOverflow) {
        AlertDialog.Builder builder = new AlertDialog.Builder(imgOverflow.getRootView().getContext());
        LayoutInflater inflater = LayoutInflater.from(imgOverflow.getRootView().getContext());
        View dialogView;
        Reusable reusable = new Reusable();
        if(commentUserId.equals(this.userId))
            dialogView = inflater.inflate(R.layout.dialog_mine_comment, null);
        else
            dialogView = inflater.inflate(R.layout.dialog_member_comment, null);
        builder.setView(dialogView);
        final AlertDialog dialog= builder.create();
        dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        dialog.show();

        Button btnSubmit, btnDelete, btnFollow, btnShare;
        btnSubmit = dialog.findViewById(R.id.btnSubmit); btnSubmit.setVisibility(View.GONE);
        btnDelete = dialog.findViewById(R.id.btnDelete);
        if(!commentUserId.equals(this.userId))
            btnDelete.setVisibility(View.GONE);
        btnFollow = dialog.findViewById(R.id.btnFollow);
        btnShare = dialog.findViewById(R.id.btnShare);

        if(UserNetwork.getFollowing()==null)
            btnFollow.setVisibility(View.GONE);
        else
            btnFollow.setText(UserNetwork.getFollowing().contains(commentUserId)? "UNFOLLOW": "FOLLOW");

        btnShare.setOnClickListener(v -> {
            reusable.shareComment(btnShare.getContext(), model.getUsername(), model.getContent());
            dialog.cancel();
        });

        btnFollow.setOnClickListener(v -> {
            if(!Reusable.getNetworkAvailability(context)){
                Snackbar.make(btnFollow, "No Internet connection", Snackbar.LENGTH_SHORT).show();
                dialog.cancel();
                return;
            }
            if (userId.equals(Calculations.GUEST)) {
                loginPrompt(btnFollow);
                return;
            }
            if(btnFollow.getText().equals("FOLLOW")){
                calculations.followMember(imgOverflow, userId, commentUserId, anchorSnackbar);
            }
            else
                unfollowPrompt(imgOverflow, commentUserId, model.getUsername());
            dialog.cancel();
        });

        btnDelete.setOnClickListener(v -> {
            deleteComment(postId, imgOverflow);
            dialog.cancel();
        });
    }

    private void unfollowPrompt(ImageView imgOverflow, String userID, String username){
        androidx.appcompat.app.AlertDialog.Builder builder = new androidx.appcompat.app.AlertDialog.Builder(context, R.style.CustomMaterialAlertDialog);
        builder.setMessage(String.format("Do you want to unfollow %s?", username))
                .setTitle("Unfollow")
                .setNegativeButton("No", (dialogInterface, i) -> {
                    //do nothing
                })
                .setPositiveButton("Yes", (dialogInterface, i) -> calculations.unfollowMember(imgOverflow, userId, userID, anchorSnackbar))
                .show();
    }

    private void deleteComment(String postId, View imgOverflow){
        final DocumentReference postPath =  FirebaseUtil.getFirebaseFirestore().collection("posts").document(mainPostId);
        FirebaseUtil.getFirebaseFirestore().runTransaction((Transaction.Function<Void>) transaction -> {
            DocumentSnapshot snapshot = transaction.get(postPath);
            //Check if post exist first
            if(!snapshot.exists()){
                Log.i(TAG, "apply: snapshot is empty");
                return null;
            }

            Map<String, Object> updates = new HashMap<>();
            final long commentCount = snapshot.toObject(Post.class).getCommentsCount() -1; //Retrieve commentCount stat
            updates.put("commentsCount", Math.max(0,commentCount));
            transaction.update(postPath, updates);
            return null;
        }).addOnSuccessListener(new OnSuccessListener<Void>() {
            @Override
            public void onSuccess(Void aVoid) {
                FirebaseUtil.getFirebaseFirestore().collection("comments").document(postId).delete()
                        .addOnSuccessListener(new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                Snackbar.make(imgOverflow, "Comment deleted", Snackbar.LENGTH_SHORT).show();
                            }
                        }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Snackbar.make(imgOverflow, "Something went wrong", Snackbar.LENGTH_SHORT).show();
                    }
                });
            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {
                Snackbar.make(imgOverflow, "Something went wrong", Snackbar.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public CommentHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.comment_view, parent, false);
        return new CommentHolder(view);
    }

    class CommentHolder extends RecyclerView.ViewHolder {
        CircleImageView imgDp;
        LinearLayout lnrContainer;
        TextView mComment;
        TextView mUsername;
        TextView mTime;
        TextView mLikes, mDislikes;
        ImageView imgOverflow;
        ImageView imgLikes, imgDislike, imgReply;
        CommentHolder(View itemView) {
            super(itemView);
            lnrContainer = itemView.findViewById(R.id.container_post);
            imgDp = itemView.findViewById(R.id.imgDp);
            mComment = itemView.findViewById(R.id.txtPost);
            mUsername = itemView.findViewById(R.id.txtUsername);
            mTime = itemView.findViewById(R.id.txtTime);
            mLikes = itemView.findViewById(R.id.txtLike);
            mDislikes = itemView.findViewById(R.id.txtDislike);

            imgLikes = itemView.findViewById(R.id.imgLike);
            imgDislike = itemView.findViewById(R.id.imgDislike);
            imgReply = itemView.findViewById(R.id.imgReply);
            imgOverflow = itemView.findViewById(R.id.imgOverflow);
        }
    }
}

File: app/src/main/java/com/sqube/tipshub/adapters/CommentAdapter.kt
package com.sqube.tipshub.adapters

import com.sqube.tipshub.adapters.CommentAdapter.CommentHolder
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.signature.ObjectKey
import com.firebase.ui.firestore.FirestoreRecyclerAdapter
import com.firebase.ui.firestore.FirestoreRecyclerOptions
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.Transaction
import com.google.firebase.storage.FirebaseStorage
import com.sqube.tipshub.activities.LoginActivity
import com.sqube.tipshub.activities.MemberProfileActivity
import com.sqube.tipshub.activities.MyProfileActivity
import com.sqube.tipshub.R
import com.sqube.tipshub.databinding.CommentViewBinding
import com.sqube.tipshub.models.Comment
import com.sqube.tipshub.models.Post
import com.sqube.tipshub.models.SnapId
import com.sqube.tipshub.models.UserNetwork
import com.sqube.tipshub.services.GlideApp
import com.sqube.tipshub.utils.Calculations
import com.sqube.tipshub.utils.FirebaseUtil.firebaseFirestore
import com.sqube.tipshub.utils.GUEST
import com.sqube.tipshub.utils.Reusable
import com.sqube.tipshub.utils.Reusable.Companion.applyLinkfy
import com.sqube.tipshub.utils.Reusable.Companion.getNetworkAvailability
import com.sqube.tipshub.utils.Reusable.Companion.getPlaceholderImage
import com.sqube.tipshub.utils.Reusable.Companion.getTime
import com.sqube.tipshub.utils.Reusable.Companion.signature
import java.util.*
import kotlin.math.min

class CommentAdapter(private val mainPostId: String, query: Query?, userID: String, val activity: Activity, val context: Context) : FirestoreRecyclerAdapter<Comment, CommentHolder>(FirestoreRecyclerOptions.Builder<Comment>()
        .setQuery((query)!!, Comment::class.java)
        .build()) {
    private val tag = "CommentAdapter"
    private var userId: String = userID
    private val calculations = Calculations(context)
    private val storageReference = FirebaseStorage.getInstance().reference.child("profile_images")
    val repliedList = ArrayList<SnapId>()

    fun setUserId(userId: String) {
        this.userId = userId
    }

    fun resetRepliesList() {
        repliedList.clear()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onBindViewHolder(holder: CommentHolder, position: Int, model: Comment) {
        val binding = holder.binding
        val postId = snapshots.getSnapshot(position).id
        if (model.isFlag)
            binding.containerPost.setBackgroundColor(context.resources.getColor(R.color.comment_flagged))
        else
            binding.containerPost.setBackgroundColor(context.resources.getColor(R.color.comment_bg))
        binding.imgReply.visibility = if ((model.userId == userId)) View.GONE else View.VISIBLE

        with(model){
            binding.txtUsername.text = username
            binding.txtPost.text = content
            applyLinkfy(context, content, binding.txtPost)
            binding.txtTime.text = getTime(time)
            binding.imgLike.setColorFilter(if (likes.contains(userId)) context.resources.getColor(R.color.likeGold) else context.resources.getColor(R.color.likeGrey))
            binding.imgDislike.setColorFilter(if (dislikes.contains(userId)) context.resources.getColor(R.color.likeGold) else context.resources.getColor(R.color.likeGrey))
            binding.txtLike.text = if (likesCount == 0L) "" else likesCount.toString()
            binding.txtDislike.text = if (dislikesCount == 0L) "" else dislikesCount.toString()
            GlideApp.with(context)
                    .load(storageReference.child(userId))
                    .placeholder(R.drawable.dummy)
                    .error(getPlaceholderImage(userId[0]))
                    .signature(ObjectKey(userId + "_" + signature))
                    .into(binding.imgDp)
        }

        //listen to dp click and open user profile
        binding.imgDp.setOnClickListener {
            if ((model.userId == userId)) {
                binding.imgDp.context.startActivity(Intent(context, MyProfileActivity::class.java))
            } else {
                val intent = Intent(context, MemberProfileActivity::class.java)
                intent.putExtra("userId", model.userId)
                binding.imgDp.context.startActivity(intent)
            }
        }

        //listen to username click and open user profile
        binding.txtPost.setOnClickListener {
            if ((model.userId == userId))
                binding.txtPost.context.startActivity(Intent(context, MyProfileActivity::class.java))
            else {
                val intent = Intent(context, MemberProfileActivity::class.java)
                intent.putExtra("userId", model.userId)
                binding.txtUsername.context.startActivity(intent)
            }
        }
        binding.imgLike.setOnClickListener {
            if ((userId == GUEST)) { loginPrompt(binding.imgLike)
                return@setOnClickListener
            }
            if (model.dislikes.contains(userId)) {
                binding.imgLike.setColorFilter(context.resources.getColor(R.color.likeGold))
                binding.imgDislike.setColorFilter(context.resources.getColor(R.color.likeGrey))
                binding.txtLike.text = (model.likesCount + 1).toString()
                binding.txtDislike.text = if (model.dislikesCount - 1 > 0) (model.dislikesCount - 1).toString() else ""
            }
            else {
                if (model.likes.contains(userId)) {
                    binding.imgLike.setColorFilter(context.resources.getColor(R.color.likeGrey))
                    binding.txtLike.text = if (model.likesCount - 1 > 0) (model.likesCount - 1).toString() else ""
                } else {
                    binding.imgLike.setColorFilter(context.resources.getColor(R.color.likeGold))
                    binding.txtLike.text = (model.likesCount + 1).toString()
                }
            }
            val substring: String = model.content.substring(0, Math.min(model.content.length, 90))
            calculations.onCommentLike(userId, model.userId, postId, mainPostId, substring)
        }
        binding.imgReply.setOnClickListener {
            if ((userId == GUEST)) { loginPrompt(binding.imgReply)
                return@setOnClickListener
            }
            val edtComment: EditText = activity.findViewById(R.id.edtComment)
            val comment: String = edtComment.text.toString()
            if (comment.isEmpty()) edtComment.setText(String.format("@%s ", model.username)) else edtComment.setText(String.format("%s @%s ", comment.trim { it <= ' ' }, model.username))
            edtComment.setSelection(edtComment.text.length)
            repliedList.add(SnapId(model.userId, model.username))
        }
        binding.imgDislike.setOnClickListener {
            if ((userId == GUEST)) { loginPrompt(it)
                return@setOnClickListener
            }
            if (model.likes.contains(userId)) {
                binding.imgLike.setColorFilter(context.resources.getColor(R.color.likeGrey))
                binding.imgDislike.setColorFilter(context.resources.getColor(R.color.likeGold))
                binding.txtLike.text = if (model.likesCount - 1 > 0) (model.likesCount - 1).toString() else ""
                binding.txtDislike.text = (model.dislikesCount + 1).toString()
            }
            else {
                if (model.dislikes.contains(userId)) {
                    binding.imgDislike.setColorFilter(context.resources.getColor(R.color.likeGrey))
                    binding.txtDislike.text = if (model.dislikesCount - 1 > 0) (model.dislikesCount - 1).toString() else ""
                } else {
                    binding.imgDislike.setColorFilter(context.resources.getColor(R.color.likeGold))
                    binding.txtDislike.text = (model.dislikesCount + 1).toString()
                }
            }
            val substring: String = model.content.substring(0, min(model.content.length, 90))
            calculations.onCommentDislike(userId, model.userId, postId, mainPostId, substring)
        }
        binding.imgOverflow.setOnClickListener { displayOverflow(model, model.userId, postId, it) }
    }

    private fun loginPrompt(view: View) {
        val builder = AlertDialog.Builder(view.rootView.context,
                R.style.CustomMaterialAlertDialog)
        builder.setMessage("You have to login first")
                .setNegativeButton("Cancel") { _: DialogInterface?, i: Int -> }
                .setPositiveButton("Login") { _: DialogInterface?, i: Int -> view.context.startActivity(Intent(view.context, LoginActivity::class.java)) }
                .show()
    }

    private fun displayOverflow(model: Comment, commentUserId: String, postId: String, imgOverflow: View) {
        val builder = android.app.AlertDialog.Builder(imgOverflow.rootView.context)
        val inflater = LayoutInflater.from(imgOverflow.rootView.context)
        val dialogView: View = if ((commentUserId == userId)) inflater.inflate(R.layout.dialog_mine_comment, null)
                else inflater.inflate(R.layout.dialog_member_comment, null)
        builder.setView(dialogView)
        val dialog = builder.create()
        val reusable = Reusable()
        dialog.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        dialog.show()
        val btnSubmit: Button = dialog.findViewById(R.id.btnSubmit)
        val btnDelete: Button = dialog.findViewById(R.id.btnDelete)
        val btnFollow: Button = dialog.findViewById(R.id.btnFollow)
        val btnShare: Button = dialog.findViewById(R.id.btnShare)
        btnSubmit.visibility = View.GONE
        if (commentUserId != userId) btnDelete.visibility = View.GONE
        if (UserNetwork.following == null)
            btnFollow.visibility = View.GONE
        else
            btnFollow.text = if (UserNetwork.following.contains(commentUserId)) "UNFOLLOW" else "FOLLOW"
        btnShare.setOnClickListener { reusable.shareComment(btnShare.context, model.username, model.content)
            dialog.cancel()
        }
        btnFollow.setOnClickListener {
            if (!getNetworkAvailability(context)) {
                Snackbar.make(btnFollow, "No Internet connection", Snackbar.LENGTH_SHORT).show()
                dialog.cancel()
                return@setOnClickListener
            }
            if ((userId == GUEST)) {
                loginPrompt(btnFollow)
                return@setOnClickListener
            }
            if ((btnFollow.text == "FOLLOW")) {
                calculations.followMember(imgOverflow, userId, commentUserId)
            } else unfollowPrompt(imgOverflow, commentUserId, model.username)
            dialog.cancel()
        }
        btnDelete.setOnClickListener { deleteComment(postId, imgOverflow)
            dialog.cancel()
        }
    }

    private fun unfollowPrompt(imgOverflow: View, userID: String, username: String) {
        val builder = AlertDialog.Builder(context, R.style.CustomMaterialAlertDialog)
        builder.setMessage(String.format("Do you want to unfollow %s?", username))
                .setTitle("Unfollow")
                .setNegativeButton("No") { _: DialogInterface?, _: Int -> }
                .setPositiveButton("Yes") { _: DialogInterface?, _: Int -> calculations.unfollowMember(imgOverflow, userId, userID) }
                .show()
    }

    private fun deleteComment(postId: String, imgOverflow: View) {
        val postPath = firebaseFirestore!!.collection("posts").document(mainPostId)
        firebaseFirestore!!.runTransaction { transaction: Transaction ->
            val snapshot: DocumentSnapshot = transaction.get(postPath)
            //Check if post exist first
            if (snapshot.exists()) {
                val updates: MutableMap<String, Any> = HashMap()
                val commentCount: Long = snapshot.toObject(Post::class.java)!!.commentsCount - 1 //Retrieve commentCount stat
                updates.put("commentsCount", Math.max(0, commentCount))
                transaction.update(postPath, updates)
                Log.i(tag, "apply: snapshot is empty")
            }
            null
        }.addOnSuccessListener {
            firebaseFirestore!!.collection("comments").document(postId).delete()
                    .addOnSuccessListener { Snackbar.make(imgOverflow, "Comment deleted", Snackbar.LENGTH_SHORT).show() }
                    .addOnFailureListener { Snackbar.make(imgOverflow, "Something went wrong", Snackbar.LENGTH_SHORT).show() }
        }.addOnFailureListener { Snackbar.make(imgOverflow, "Something went wrong", Snackbar.LENGTH_SHORT).show() }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CommentHolder {
        val binding = CommentViewBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return CommentHolder(binding)
    }

    inner class CommentHolder(val binding: CommentViewBinding) : RecyclerView.ViewHolder(binding.root)
}